File: Roadmap/01 - OPERADORES Y ESTRUCTURAS DE CONTROL/javascript/Roadmap/10 - EXCEPCIONES/python/garos01.py
a = 5
b = 0
try:
    # Intenta ejecutar el código que podría generar un error
    resultado = a / b  # Intentando dividir por cero
except ZeroDivisionError as error:
    # Captura la excepción específica y maneja el error
    print(f"Error: {error}")
else:
    # Este bloque se ejecuta si no hay excepciones
    print(f"La división fue exitosa: {resultado}")
finally:
    # Este bloque se ejecuta siempre, independientemente de si hubo una excepción o no
    print("Fin del manejo de excepciones")


lista = [1, 2, 3]
try:
    # Intenta acceder a un índice no existente de una lista
    elemento = lista[4]  # Intentando acceder a un índice no existente
except IndexError as error:
    # Captura la excepción específica y maneja el error
    print(f"Error: {error}")
else:
    # Este bloque se ejecuta si no hay excepciones
    print(f"Elemento obtenido: {elemento}")
finally:
    # Este bloque se ejecuta siempre, independientemente de si hubo una excepción o no
    print("Fin del manejo de excepciones")


# Ejercicio extra


class MiExcepcionPersonalizada(Exception):
    pass


def procesar_parametros(parametro):
    try:
        if parametro == 0:
            raise ValueError("El parámetro no puede ser cero.")
        elif parametro < 0:
            raise ValueError("El parámetro no puede ser negativo.")
        elif parametro > 100:
            raise MiExcepcionPersonalizada("El parámetro es demasiado grande.")
        else:
            print("El parámetro es válido.")
    except ValueError as error:
        print("Error:", error)
    except MiExcepcionPersonalizada as error_personalizado:
        print("Error personalizado:", error_personalizado)
    else:
        print("No se ha producido ningún error.")
    finally:
        print("La ejecución ha finalizado.")


try:
    procesar_parametros(50)
    print()
    procesar_parametros(0)
    print()
    procesar_parametros(-5)
    print()
    procesar_parametros(200)
except Exception as error:
    print(f"Error capturado desde el lugar donde se llama a la función: {error}")
finally:
    print("Fin del programa.")

File: Roadmap/10 - EXCEPCIONES/javascript/garos01.js
let a = 5;
let b = 0;

try {
  // Intenta ejecutar el código que podría generar un error
  let resultado = a / b; // Intentando dividir por cero
  console.log(`La división fue exitosa: ${resultado}`);
} catch (error) {
  // Captura la excepción específica y maneja el error
  console.error(`Error: ${error}`);
} finally {
  // Este bloque se ejecuta siempre, independientemente de si hubo una excepción o no
  console.log("Fin del manejo de excepciones");
}

let lista = [1, 2, 3];

try {
  // Intenta acceder a un índice no existente de una lista
  let elemento = lista[4]; // Intentando acceder a un índice no existente
  console.log(`Elemento obtenido: ${elemento}`);
} catch (error) {
  // Captura la excepción específica y maneja el error
  console.error(`Error: ${error}`);
} finally {
  // Este bloque se ejecuta siempre, independientemente de si hubo una excepción o no
  console.log("Fin del manejo de excepciones");
}

// Ejercicio extra

class MiErrorPersonalizado extends Error {
  constructor(message) {
    super(message);
    this.name = "MiErrorPersonalizado";
  }
}

function procesarParametros(parametro) {
  try {
    if (typeof parametro !== "number") {
      throw new TypeError("El parámetro debe ser un número");
    }

    if (parametro === 0) {
      throw new Error("El parámetro no puede ser cero");
    }

    if (parametro < 0) {
      throw new MiErrorPersonalizado("El parámetro no puede ser negativo");
    }

    console.log("El parámetro es válido:", parametro);
  } catch (error) {
    console.error("Tipo de error:", error.name);
    console.error("Mensaje de error:", error.message);
  } finally {
    console.log("La ejecución ha finalizado");
  }
}

try {
  procesarParametros("cadena");
  procesarParametros(0);
  procesarParametros(-5);
  procesarParametros(10);
} catch (error) {
  console.error(`Se ha producido un error fuera de la función: ${error}`);
}
